crates/interpreter/src/interpreter/shared_memory.rs

```diff
@@ -0,0 +1,286 @@
+use revm_primitives::U256;
+
+use crate::alloc::vec;
+use crate::alloc::{boxed::Box, slice, vec::Vec};
+use core::{
+    cmp::min,
+    fmt,
+    ops::{BitAnd, Not},
+};
+
+/// A sequential memory shared between calls, which uses
+/// a `Vec` for internal representation.
+/// A [SharedMemory] instance should always be obtained using
+/// the `new` static method to ensure memory safety.
+pub struct SharedMemory {
+    /// Shared buffer
+    data: Box<[u8]>,
+    /// Memory checkpoints for each depth
+    checkpoints: Vec<usize>,
+    /// Raw pointer used for the portion of memory used
+    /// by the current context
+    current_ptr: *mut u8,
+    /// How much memory has been used in the current context
+    current_len: usize,
+    /// Amount of memory left for assignment
+    pub limit: usize,
+}
+
+impl fmt::Debug for SharedMemory {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("SharedMemory")
+            .field(
+                "current_slice",
+                &crate::primitives::hex::encode(self.current_slice()),
+            )
+            .finish()
+    }
+}
+
+impl SharedMemory {
+    /// Calculates memory allocation upper bound using
+    /// https://2π.com/22/eth-max-mem
+    #[inline]
+    pub fn calculate_upper_bound(gas_limit: u64) -> u64 {
+        4096 * sqrt(2u64.checked_mul(gas_limit).unwrap_or(u64::MAX))
```

Comment:
```suggestion
        4096 * sqrt(2u64.saturating_mul(gas_limit))
```

Reply:
My bad that I don't know Rust very well yet